### ISSUE TITLE
[SOLVED] ACER Aspire-VN7-792G with Skylake-H - Sound starts working but still no D-Mic

### ISSUE THREAD

**Issue by Gnoenk** [thesofproject/avs-topology-xml#27]
Dear @crojewsk-intel

firstly I want to thank you so much for your advice and work in resolving the initial D-Mic issue and subsequent applicability of the new AVS Driver for the Intel Skylake-H platform! As I am only an amateur I needed a bit of time to do tests and research and responding back to you and I highly appreciate your patience.

As the foregoing issue https://github.com/thesofproject/avs-topology-xml/issues/24 was closed before I could submit my update I do not know to what extent you noted that status reported there. On top of that I did further efforts in the meantime and can report a number of sucesses! However there are still open issues thus I open this issue as per your suggestion.

Current Status with the AVS-driver on my machine now is:
+ Sound from internal speaker works
    + inserting stereo headphones: automatic switch to this device and sound work
    + disconnecting headphones: switch back to internal audio speakers works
+ inserting an external mic/speaker device via USB
    + gets displayed as new option in `pavucontrol`.
    + selecting it:  sounds work
    + also the microphone from this device is working
+ connecting a TV via HDMI cable
    + sound works

- but overall output volume-level is very low. Putting it to max volume results in still low output.
- build in DMIC: does not work

What have I done?

STEP 1: Update of System Configuration
As described in more details in my last post in https://github.com/thesofproject/avs-topology-xml/issues/24 I have now the new configuration
  -Kernel 6.8.0.39
  -Alsa-Lib 1.2.11-1build2
  -Alsa-Utils 1.2.11
  -replaced UCM tplg.bin files
  -added UCM config files

and the system successfully does playback of white noise via the `aplay` command (tested as suggested by you in two versions: with and without subformat option).

Unfortunately under normal operation there was still no sound and no microphone functionality.`pavucontrol` did not show any input or output devices.

STEP 2: 
As noted in the previous post "to our knowledge MINT22 has switched from pulseaudio to pipewire with wireplumber" - so I did futher research and testing and I learned about `alsa-info.sh` which is very useful and `pactl`.
Despite the fact that alsa-info showed only pipe-wire as running sound server I tried `pactl`.
Applying   `pactl list cards` and `pactl list sinks` showed that the audio-card had no activated profile. Using
  `pactl set-card-profile 47 pro-audio`
did activate the card resulting in the status as summarized in the beginning.

OPEN issues:
Two issues remain:
- overall sound output level is low
- D-MIC not working.

By rerunning the `pactl list` commands I collected info about the sources/sinks status , see 
[pactl list sources and sinks.txt](https://github.com/user-attachments/files/17174598/pactl.list.sources.and.sinks.txt) .

By using the sript `alsa-info` I have created this output 
[alsa-info.txt.11xo06vqsE.txt](https://github.com/user-attachments/files/17174600/alsa-info.txt.11xo06vqsE.txt) .
Here I noted in the Section "   !!ALSA/HDA dmesg   "   that the dmesg output contains DMIC related statements and I could spot lines reporting Dmic errors followed by this note
" snd_soc_avs 0000:00:1f.3: no DMIC ep_blob found "

Hope that this provides useful information for further analysis and the we will be able to identify the last missing pieces of the puzzle!

Finally one more question: what kind of card is avsprobemb ?

Thank you again - and kind regards!

**Comment by d2cmath** [comment]
From your `alsa-info.txt`, your both your physical microphone and DMIC volume is 0. You can turn them up by running `alsamixer` command in the terminal (F6 to choose device, Tab to switch between Playback/Input, and arrow keys to choose and configure the values or using number keys). You can also turn up your Master Volume of Playback with this.

The message you see spammed in dmesg is the glory of snd_soc_avs dynamic management, nothing serious; avsprobemb is probably a dummy device created when nothing is probed at the system start. Your main sound card is Realtek ACL255, which is device no. 2 and is the default. I was surprised that it did not work out of the box with the auto mode of snd_intel_dspcfg and recent kernels.

**Comment by Gnoenk** [comment]
Dear @d2cmath

thank you for your reply and comments!

I have used `alsamixer` to change the settings following your advice
Using F6 I did select the cards: avs_dmic and hdaudioB0D0 and adjusted the controls being zero upwards.

Result is that the sound level is now working fine!
Unfortunately despite having adjusted the level for the D-Mic I still see no effect: D-Mic is still not working.

I noted that alsamixer displays for the card hdaudioB0D0 the chip ALC255 but for the card avs_dmic the chip has no entry.
I have created a new report by `alsa-info` [alsa-info.txt.PaUB9vgw7U.txt](https://github.com/user-attachments/files/17247651/alsa-info.txt.PaUB9vgw7U.txt) . Hope you can identify the modified settings here. 
Question: which section does show these values?

So (the initial) one issue remains:
- D-MIC not working.

Do the DMESG notifications about the missing blob and the other DMIC related errors provide more hints?
I am very grateful for any ideas what to check next.

Note: the card avsprobemb can be selected in alsa-mixer but no controls to be adjusted exist. So it seems that your guess is right.

**Comment by d2cmath** [comment]
Strange, your internal microphone should work now. According to alsa-info, its volume is now set to 82%, it is said in the section starting with `Mixer controls for card avsdmic`. 

Can you run the following command, make some noise near your laptop to see if it records something on the screen?

`arecord -D hw:2,2 -f S16_LE -c 2 -r 48000 -vv > /dev/null`

Note: `hw:2,2` here is your DMIC device according to also-info (section `List of CAPTURE Hardware Devices`) but this can change in each boot, so if the command complains about non-existing file then run `arecord -l` to see which card and device is DMIC now. It may complain about the format in the `-f S16_LE` part then follow the suggestion to change to a different format (eg. `S32_LE`), etc.

**Comment by Gnoenk** [comment]
Dear @d2cmath


thank you for your suggestions!

I have run the commands with the following results: 

Firstly I checked the card numbers.
~~~ console
~$ LANG=C arecord -l
**** List of CAPTURE Hardware Devices ****
card 1: hdaudioB0D0 [hdaudioB0D0], device 1: HDAudio Analog (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 2: avsdmic [avs_dmic], device 2: Digital Microphone (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
~~~

Then I modified this command slightly to get the output into a file.
~~~ console
~$ arecord -D hw:2,2 -f S16_LE -c 2 -r 48000 -vv > recording.pcm 
Hardware PCM card 2 'avs_dmic' device 2 subdevice 0
Its setup is:
  stream       : CAPTURE
  access       : RW_INTERLEAVED
  format       : S16_LE
  subformat    : STD
  channels     : 2
  rate         : 48000
  exact rate   : 48000 (48000/1)
  msbits       : 16
  buffer_size  : 24000
  period_size  : 6000
  period_time  : 125000
  tstamp_mode  : ENABLE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 6000
  period_event : 0
  start_threshold  : 1
  stop_threshold   : 24000
  silence_threshold: 0
  silence_size : 0
  boundary     : 6755399441055744000
  appl_ptr     : 0
  hw_ptr       : 0
Aufnahme: WAVE 'stdin' : Signed 16 bit Little Endian, Rate: 48000 Hz, stereo
##+                                                | 03%^CAbbruch durch Signal Unterbrechung ...
~~~
During the recording I was  also shouting but the percent indicator did not exceed 05%.

Then I played the output file and could hear what I said - but it was very very silent.

~~~ console
~$ aplay -f S16_LE -c 2 -r 48000 recording.pcm 
Wiedergabe: WAVE 'recording.pcm' : Signed 16 bit Little Endian, Rate: 48000 Hz, stereo
~~~ 

Finally I checked the file content to be able to display the sound output to you,


~~~ console
~$ hexdump -C -n 512 recording.pcm 
00000000  52 49 46 46 24 00 00 80  57 41 56 45 66 6d 74 20  |RIFF$...WAVEfmt |
00000010  10 00 00 00 01 00 02 00  80 bb 00 00 00 ee 02 00  |................|
00000020  04 00 10 00 64 61 74 61  00 00 00 80 00 00 00 00  |....data........|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000000b0  00 00 00 00 00 00 00 00  ff ff 00 00 0a 00 2a 00  |..............*.|
000000c0  bc 00 c8 01 2b 03 15 05  fe 04 c4 04 d0 03 a1 02  |....+...........|
000000d0  e0 03 73 04 6f 04 c3 03  c0 03 79 03 4d 04 4e 04  |..s.o.....y.M.N.|
000000e0  f7 03 4f 03 16 04 32 04  24 04 90 03 f6 03 e9 03  |..O...2.$.......|
000000f0  34 04 d1 03 f5 03 b5 03  29 04 f3 03 05 04 a5 03  |4.......).......|
00000100  1d 04 f9 03 0e 04 af 03  13 04 e9 03 19 04 c1 03  |................|
00000110  0c 04 d2 03 1c 04 d2 03  0d 04 c8 03 17 04 d2 03  |................|
00000120  11 04 ca 03 17 04 d4 03  14 04 cd 03 11 04 d3 03  |................|
00000130  15 04 cd 03 10 04 c8 03  0e 04 c8 03 09 04 c9 03  |................|
00000140  10 04 c8 03 10 04 ca 03  0b 04 ca 03 11 04 ca 03  |................|
00000150  0b 04 cc 03 0f 04 c9 03  10 04 c9 03 0f 04 cb 03  |................|
00000160  0d 04 c8 03 0d 04 c9 03  12 04 c9 03 0b 04 ca 03  |................|
00000170  0d 04 ca 03 0d 04 ca 03  0e 04 c8 03 11 04 ca 03  |................|
00000180  0c 04 cb 03 0d 04 c6 03  0f 04 c9 03 0e 04 cb 03  |................|
00000190  0c 04 c9 03 0e 04 c9 03  0f 04 cc 03 0d 04 ca 03  |................|
000001a0  0d 04 c7 03 0c 04 c9 03  0f 04 c9 03 0f 04 c9 03  |................|

~~~

I am surprised - indeed the D-Mic is getting activated by the `arecord` command!

But the sound detection level is very very low. And applications seem to not be able to connect to this input channel at all.
Now I am curious: what to check next?

One more thought:` pactl list sources` displays only two sources - both are related to alsa_input.platform-avs_hdaudio (this is listed in the textfile attached in the first post). The avs_dmic is not mentioned here - is this hinting to something?
And `pactl info`shows:
~~~ console
~$ LANG=C pactl info
Server String: /run/user/1000/pulse/native
Library Protocol Version: 35
Server Protocol Version: 35
Is Local: yes
Client Index: 122
Tile Size: 65472
User Name: g
Host Name: g-Aspire-VN7-792G
Server Name: PulseAudio (on PipeWire 1.0.5)
Server Version: 15.0.0
Default Sample Specification: float32le 2ch 48000Hz
Default Channel Map: front-left,front-right
Default Sink: alsa_output.platform-avs_hdaudio.0.pro-output-1
Default Source: alsa_input.platform-avs_hdaudio.0.pro-input-1
Cookie: d2e6:1130
~~~

**Comment by d2cmath** [comment]
The fact that the `arecord` command runs without any error (and records some signals) means that the Linux kernel recognised your built-in microphone perfectly and has successfully set up the ALSA driver. 

I am a little bit out of idea here, for now I can only see one possibility: something is physically obstructing the signal, ie. either the microphone holes are covered or dirty, or your hardware is damaged, eg. the wiring went bad. I googled your model, it seems the microphone holes are on top of your monitor (to the left of the webcam it seem, but I am not sure), can you check and clean those a little bit? Or can you confirm if they work fine on another OS to refute my hypothesis?

Note: also make sure to turn up the volume of DMIC with `alsamixer` before testing with `arecord`.

About your question about `pactl list sources`, it is normal (you will see the same result with WirePlumber control by running `wpctl status`). DMIC will not be shown in the list of sources in the sound server of PipeWire/WirePlumber if it has not yet been set up with a proper User Case Management files in ALSA. For AVS driver of DMIC, I believe the system will look for the file `/usr/share/alsa/ucm2/conf.d/avs_dmic/avs_dmic.conf`. Here is [an example](https://github.com/WeirdTreeThing/chromebook-ucm-conf/tree/main/avs/avs_dmic) from what people with Chromebook use. Your device is not a Chromebook, so you need to adapt a little bit (eg. removing the OS version checking part in HiFi.conf and replace the line below on `CaptureChannels` directly with value 2, this is also the number of working channels `-c 2` in the `arecord` command that you has run).

**Comment by Gnoenk** [comment]
Your research is correct. The physical location of the two microphones is left to the webcam. Both holes are clean.
The notebook is configured as a dualboot machine with Windows 10 and Linux Mint. Indeed the DMIC as it is works perfectly fine under Windows 10.
Using `alsamixer` I have put the DMIC level to 100% before testing it with `arecord`.

I tried the `wpctl` command.

~~~ console
$ wpctl status
PipeWire 'pipewire-0' [1.0.5, g@g-Aspire-VN7-792G, cookie:748856243]
 └─ Clients:
        32. pipewire                            [1.0.5, g@g-Aspire-VN7-792G, pid:1787]
        34. WirePlumber                         [1.0.5, g@g-Aspire-VN7-792G, pid:1786]
        35. WirePlumber [export]                [1.0.5, g@g-Aspire-VN7-792G, pid:1786]
        57. Cinnamon Volume Control Media Keys  [1.0.5, g@g-Aspire-VN7-792G, pid:1937]
        58. Cinnamon Volume Control             [1.0.5, g@g-Aspire-VN7-792G, pid:2097]
        59. Blueman                             [1.0.5, g@g-Aspire-VN7-792G, pid:2144]
        60. Muffin                              [1.0.5, g@g-Aspire-VN7-792G, pid:2097]
        61. xdg-desktop-portal                  [1.0.5, g@g-Aspire-VN7-792G, pid:2934]
        63. wpctl                               [1.0.5, g@g-Aspire-VN7-792G, pid:3775]

Audio
 ├─ Devices:
 │      44. Built-in Audio                      [alsa]
 │      45. Built-in Audio                      [alsa]
 │      46. Built-in Audio                      [alsa]
 │      47. Compress-Offload device (ALSA card 1) [alsa:compressed]
 │  
 ├─ Sinks:
 │  *   54. Built-in Audio Pro                  [vol: 1.00]
 │  
 ├─ Sink endpoints:
 │  
 ├─ Sources:
 │  *   55. Built-in Audio Pro                  [vol: 1.00]
 │  
 ├─ Source endpoints:
 │  
 └─ Streams:

Video
 ├─ Devices:
 │      42. HD WebCam                           [v4l2]
 │      43. HD WebCam                           [v4l2]
 │  
 ├─ Sinks:
 │  
 ├─ Sink endpoints:
 │  
 ├─ Sources:
 │  *   48. HD WebCam (V4L2)                   
 │  
 ├─ Source endpoints:
 │  
 └─ Streams:


Settings
 └─ Default Configured Node Names:
         0. Audio/Sink    alsa_output.usb-0b0e_Jabra_SPEAK_510_USB_745C4BFA1D92021900-00.analog-stereo
         1. Audio/Source  alsa_input.platform-avs_hdaudio.0.pro-input-1
~~~

Then I looked into`/usr/share/alsa/ucm2/conf.d/avs_dmic`. Here are six files listed but none has the name `avs_dmic.conf`. However it does contain one file named `/usr/share/alsa/ucm2/conf.d/avs_dmic/Acer-Lars-1.0.conf`. The others are called `Google-Atlas-1.0.conf` or similar.

`Acer-Lars-1.0.conf`has the content:
```
Syntax` 6

SectionUseCase."HiFi" {
	File "/Intel/avs/avs_dmic/DMIC-2ch-HiFi.conf"
	Comment "Play HiFi quality Music"
}
```

I don't know whether this file is being selected by the system as it has a different name.

The folder `/usr/share/alsa/ucm2/Intel/avs/avs_dmic`does contain six files. One is named `DMIC-2ch-HiFi.conf` - the name used in the conf-file mentioned above. 
Its content is:
```
SectionDevice."Mic" {
	Comment "Digital Microphone"

	Value {
		CapturePCM "hw:${CardId},2"
		CaptureChannels 2
		CaptureMixerElem "DMIC"
	}
}
```

I understand your comment such that you recommend adding an extra `avs_dmic.conf` file and experiment with the content a bit like the example in your link?

**Comment by d2cmath** [comment]
Thanks for confirming that your hardware is not damaged and works in a different OS. So the remaining issue of DMIC comes back to either hardware setting or driver, and you need to make `arecord` to properly get the signal first as it records directly at the hardware level. Few suggestions:

1 - I noticed that your Auto Mute feature of your Realtek sound card (which is responsive for the headphone jack and Speaker) is Enabled. Can you disable it with `alsamixer` (using arrow keys) and try the test again? If this does not work, there is also the Loopback Mixing to toggle/play with around.

Usually, `arecord` will record the signal directly from the hardware that the user specifies, even though the default source is something else or it has been muted at software level (like in PipeWire/WirePlumber). However, there may be some feature embedded into the firmware/hardware of a device that prevents its functionality, like Auto Mute can try to minimise sound/noise from the environment when the headphone jack is plugged in, etc.

2 - Have you tried to see if your sound system works with Legacy driver (snd_hda_intel) rather than the new AVS driver? Like you could try Mint Wilma from Live USB (like using Ventoy) and see if the default driver (I think the system will still use the legacy driver) works better there in kernel 6.8? You can even try more a recent kernel (I think MX Linux-AHS version uses 6.10 liquorix kernel). Again you can use `arecord` and `alsamixer` to check the hardware functions first.

About the configuration of UCM end points, this is a software-level issue and can be handled later. But your device is definitely not Acer Lars nor Google Atlas (those are chromebook devices). So I do not think the files you listed will be read. To check if ALSA have set up the end points properly, you can run command `alsactl init` and see if it complains about User Case Management files or not (it will initialise the hardware and driver anyway).

Also with WirePlumber you can inspect the details of hardware devices and end-point nodes in the Audio branch of the tree shown in `wpctl status`. The command is `wpctl inspect <number>` using the <number> you see in the list. You have quite a bunch of devices connecting to your system there (like some USB speaker?), perhaps unplug them when testing DMIC.

**Comment by Gnoenk** [comment]
Thank you for your new thoughts. Before I can try them I want to report about the progress made in the last hours. 

I did create a file `/usr/share/alsa/ucm2/conf.d/avs_dmic/avs_dmic.conf` and simply copied the content of `Acer-Lars-1.0.conf` (as shown above) into the new file. Thus it refers to (the already existing) ...File "/Intel/avs/avs_dmic/DMIC-2ch-HiFi.conf"...
Then I did reboot and run `arecord` as above again. Now the recording indicator showed %-values up to 80/90 % ! Also playing the file with `aplay`now had a much higher volume level! 

A real first time on this machine :-))

Obviously this missing file made a real difference. Next I have rerun `alsa-info` and got this new report 
[alsa-info.txt.m9tdWIpD7w.txt](https://github.com/user-attachments/files/17266643/alsa-info.txt.m9tdWIpD7w.txt).  It seems to me that there are more DMIC related messages then before and they read like that something went ok when initializing the DMIC ?
However  as before the DMIC does not work with any application.


On your question 2 about trying legacy driver. 
Yes I initially used them. It took some work to configure them but I could successfully get the audio to work with exception of DMIC. Actually this is where this quest started.  Indeed I can change back easily by NOT enforcing the AVS driver as I currently do by the setting `snd_intel_dspcfg.dsp_driver=4`in `/etc/modprobe.d/alsa-base.conf` . But all the other audio seems to work fine with AVS as reported in the first post of this thread.

On your question about the USB speaker: I have tested this USB device containing both (a speaker and a microphone) successfully using AVS. But this device is disconnected and I don't know why it is listed in the report.

Now I need some time to read, understand and test your suggestions. Eventually you can guide me to test certain things first based upon the fresh information above.

This one check I could do quickly. Here it is:
~~~ console
$ wpctl inspect 44
id 44, type PipeWire:Interface:Device
    alsa.card = "1"
    alsa.card_name = "avs_dmic"
    alsa.driver_name = "snd_soc_avs_dmic"
    alsa.id = "avsdmic"
    alsa.long_card_name = "Acer-AspireVN7_792G-V1.09"
    api.acp.auto-port = "false"
    api.acp.auto-profile = "false"
    api.alsa.card = "1"
    api.alsa.card.longname = "Acer-AspireVN7_792G-V1.09"
    api.alsa.card.name = "avs_dmic"
    api.alsa.path = "hw:1"
    api.alsa.use-acp = "true"
    api.dbus.ReserveDevice1 = "Audio1"
  * client.id = "35"
  * device.api = "alsa"
    device.bus-path = "platform-avs_dmic"
  * device.description = "Built-in Audio"
    device.enum.api = "udev"
    device.form-factor = "internal"
    device.icon-name = "audio-card-analog"
  * device.name = "alsa_card.platform-avs_dmic"
  * device.nick = "avs_dmic"
    device.plugged.usec = "6270462"
    device.string = "1"
    device.subsystem = "sound"
    device.sysfs.path = "/devices/platform/avs_dmic/sound/card1"
  * factory.id = "14"
  * media.class = "Audio/Device"
    object.path = "alsa:pcm:1"
  * object.serial = "44"
~~~

**Comment by d2cmath** [comment]
OK, I am glad it helps. Anyway, If `arecord` and `alsamixer` works fine then your hardware+driver for the internal microphone (DMIC) are now working. If no application works with DMIC yet, then you now need to check and fix the sound server PipeWire/WirePlumber setup, eg. see if the source node associated with DMIC has been muted or not, or it has been set up with the right number of channels or not, etc.

Can you tell if the output of `alsactl init` has any error in it? If not then, in PipeWire PulseAudio controller/mixer you should see your internal microphone as a source node now: on Linux Mint, the mixer is accessible on the system tray with the Speaker icon, click on and choose Audio Mixer. Look at the Input Devices and select the Internal Microphone, make sure that it is not muted (usually there are three icons on top of each device: Mute Audio, Lock Channels and Set as Default; so make sure that Mute Audio is not pressed). You can also see if the node is recording any signal or not (similar to `arecord`, but here we are at the software layer).

**Comment by Gnoenk** [comment]
I just run`alsactl init` and the outcome is
~~~ console
~$ alsactl init 1
alsa-lib parser.c:261:(parse_syntax_field) Syntax field not found in avs_dmic.conf
alsa-lib main.c:1554:(snd_use_case_mgr_open) error: failed to import hw:1 use case configuration -22
Found hardware: "avs_dmic" "" "" "" ""
Hardware is initialized using a generic method
~~~
It seems that it looks into `avs_dmic.conf` but does not find what is being expected.

In the Linus Mint Audio Mixer I do see only one Audio-Input Device with the name Built-in Audio Pro. I cannot see the three Icons you mentioned. This device name  Built-in Audio Pro is also shown for Audio Output.

I now also tried to change the settings as you suggested for the  Auto Mute feature of the Realtek sound card (which is responsive for the headphone jack and Speaker). You  have written that it is Enabled. So  I used `alsamixer` and the arrow keys for the ALC255 card but could not see any Info or Field named  AutoMute which I can adjust.

**Comment by d2cmath** [comment]
Yes, ALSA is unable to load your `avs_dmic.conf` file yet because the line `Syntax <some number>` is missing for the main configuration file of the device. This is why PipeWire PulseAudio is unable to see and control the source node yet even though the hardware is working.  I think your use of `Acer-Lars-1.0.conf` is fine as it also includes to the 2-Chan configuration files in the `ucm2/Intel` folder, but you might have inserted some weird characters while copying at the line starting with `Syntax`. Please fix this and recheck.

Side note: in the more recent version of ucm at [Intel repository](https://github.com/amadeuszslawinski-intel/alsa-ucm-conf/tree/aslawinx/ucms/ucm2/conf.d/avs_dmic), the folders are re-organised and `Acer-Lars-1.0.conf` is only a symbolic link back to a new file in `ucm2/Intel/avs/avs_dmic/`. For your version, you can probably just create a symbolic link point to `Acer-Lars-1.0.conf` to avoid copying mistakes.

About `alsamixer`, have you checked all the sub-devices of ACL255? You need to use Tab key to switch between Playback and Capture, or All devices. Anyway, this step is now not very essential since you mentioned that `arecord` now works fine with the DMIC device and the signal is strong, so there is no obstruction at hardware level.

**Comment by Gnoenk** [comment]
Following your advice I have doublechecked my selfcreated file `avs_dmic.conf` and indeed I spotted an unwanted character within the first line of my copied content of `Acer-Lars-1.0.conf`.  You can still see it my respective post above. I believe me having it added whilst doing the  formatting step for this part of the post. This character is not in the original file of course. I had "Syntax` 6" in the first line and it should be "Syntax 6".
Deleting that and doing another reboot was an easy fix - and voilà - the LinuxMint Audio Mixer now offers in the tab "Input" a "Digital Microphone Built-In Audio" as second choice next to the "Built-in Audio Pro"!

Repeating `alsactl` now shows:
~~~ console
$ alsactl init 1
Found hardware: "avs_dmic" "" "" "" ""
Hardware is initialized using a generic method
~~~

For a quick test of the webcam and DMIC I use https://webcammictest.com/. Now here the Digital Microphone is also being offered and can be selected. And now it does record sound!

Again another "first" and a great new step forward ! 

BUT - and now again comes a "but" unfortunately - the recorded voice does not sound right. Voice is very high, squeezed and there is a kind of overlay by electric noise but I can identify the words. Sorry I don't know how to describe it better. A second test with the small Linux programm Cheese shows a similar effect.

Finally I was able to find the Auto-Mute control for the ALC255 card in `Alsamixer` and tried both options when doing the live DMIC test as described above. However that made no difference.

Note: I could not yet check the new ucm files on the Intelpage you mentioned.

We have made real progress today and I am looking forward to your comments for next checks - thank you so much!

**Comment by d2cmath** [comment]
Can you check if you have the same issue when recording directly from the hardware with `arecord`? Also try  with moderate volume for DMIC when recording rather than too high values to avoid distortion and noise. If the issue still occurs there, then try to tweak the parameter of the commands (number of channels, format, bit rate) like I have mentioned in on of the earlier response (`arecord --help` to see the meaning of different parameters). 

Otherwise, if there is no problem with `arecord` then your hardware+driver work fine, but you have an issue with the audio server. This is rather out of the scope this project, you can probably find better helps at proper channels like PipeWire's bug tracker, forums, or reddit, etc. I am not familiar with PipeWire neither, perhaps other folks can help you.

Bonus: This [video](https://www.youtube.com/watch?v=HxEXMHcwtlI) explains well the sound eco-system on Linux. You can see that if something goes wrong, it can happen at different levels :grin:. But so far, I think your hardware works fine on Linux at least at the driver+hardware level :+1:.

**Comment by Gnoenk** [comment]
I did doublecheck the outcome of recording with `arecord -D hw:1,2 -f S16_LE -c 2 -r 48000 -vv > recording.pcm` and play back with `aplay -f S16_LE -c 2 -r 48000 recording.pcm` with different settings for the DMIC volume in Alsamixer.
In all cases there is NO disturbing/crackling/etc sound overlay. However the DMIC recording volume level needs to be at 100% range otherwise the playback volume level is too low. At 100% level for DMIC recording and then setting playback volume to  130 % I get a reasonable volume output level (which is still far away from being loud). That is the last "deficit" I see here and I would like to be able to adjust more - otherwise arecord and aplay work fine. Grateful for any advice on means to adjust this further.

Then I did the same test with different settings for the DMIC volume in Alsamixer with the applications mentioned above. That made no difference with respect to the disturbing noises and the strange, high and squeezed "MickeyMouse" kind of voice.

Noting your Side Note further above, I checked your link to the more recent versions of ucm at Intel repository as per your link. Indeed I had these six files listed there all installed in `/usr/share/alsa/ucm2/conf.d/avs_dmic/`, beginning with Acer-Lars-1.0.conf. All my (older) files start with "Syntax ..." and contain some code referring to the files in `/usr/share/alsa/ucm2/Intel/avs/avs_dmic/`. Now it seems the only difference is the switch to Symlinks. In my case the addition of the file named `avs_dmic.conf` made a fundamental difference. Should and can I share this observation to the owner of this repository (if yes - how) ?

So now I will start to watch the Video and do more research on Pipewire etc. 
Thanks again - I will keep you posted.

**Comment by crojewsk-intel** [comment]
Thank you for the assistance @d2cmath, much appreciated.

@Gnoenk, both me and Amadeusz are part of the audio team. I believe it would be best to share your observations here. In the end, if the userland requires an update, the team will send the necessary changes to the relevant ALSA repo here on github [1].

If the DSP Volume mixer for the avs_dmic card is already at 100% then there is not much that can be done outside of software (userland) boosting. The Digital Microphone array is not a typical codec device, it exposes no volume control over registers e.g.: I2C/SPI registers - volume control happens over DSP. If the codec side (nonexistent) and the firmware side (existing volume control) are insufficient, software is the way to go.

Side note: the avs_probe card is for debugging purpose only - helps verify continuity of an audio stream by gathering the data directly from processing module pins. Essentially, one can verify how each data chunk changes after passing through processing module in question.

\[1\]: https://github.com/alsa-project/alsa-ucm-conf/

**Comment by Gnoenk** [comment]
Dear @crojewsk-intel,

thank you for your comments! Your help and @d2cmath assistance have enabled me to do a lot of progress in my journey to get this fine Notebook fully functioning under Linux. I really appreciate your efforts!

I do not fully understand what you mean with "share your observations HERE". Do you mean to continue in this issue or do you suggest that I should open a new issue under [1] with a summary of the current status reached so far?

On the DSP Volume Mixer: this is the one shown by Alsamixer for the avs_dmic card - correct? It is indeed at 100%. 
So you mean that any further adjustments needs to be done in extra software - correct?

What can I do next to provide usable input for further steps?

**Comment by Gnoenk** [comment]
Dear @crojewsk-intel,
dear @d2cmath,

I am delighted to be able to report to you today about my SUCCESS-ful completion of our joint work to get the DMIC working on my notebook!

During the last weeks I learned more about Pipewire and Wireplumber and experimented with commands like `pw-top`and the `arecord` / `aplay` counterparts `pw-record` / `pw-play` etc. which are explained in <https://docs.pipewire.org/page_man_pipewire_1.html> . Furthermore I changed parameters using `alsa-mixer`to adjust sound-volume (all set around 100%). Also `pa-info`was useful.

To make a long story relatively short:
I could observe via `pw-top`the actual use of the audio-format per application. 
One observation was that Pipewire mainly used S32LE and 48000.
First I did suspect a mismatch of the sample-rates being the cause of the "Mickey-Mouse" audio. But this was not the case. Then I suspected a mismatch of the audio-formats. Experiments with `pw-record`/`pw-play` were not successful - the MickeyMouse Voice was always there - irrespective of any audio-format option specified.

As the AVS driver is part of the kernel and ALSA is the next layer upon which Pipewire works, I went one level lower to test below pipewire. Suspicion was that Pipewire might receive a "bad" file/format before starting any work with it. 

So I tested again with `arecord`:

`LC_ALL=C arecord -Dhw:3,2 -c2 -r48000 -fS32_LE -vvv recording.pcm` and `aplay recording.pcm`
was resulting again in a Mickey-Mouse voice.

Then I remembered your earlier advice, dear @crojewsk, which contained S16_LE as a command parameter for `arecord`.

So next test with
`LC_ALL=C arecord -Dhw:3,2 -c2 -r48000 -fS16_LE -vvv recording2.pcm` and `aplay recording2.pcm`
resulted in a normal voice at an acceptable output volume.

So it seems that the built-in DMIC does only work correctly with S16_LE. 
This fact was not obvious to me. Some more search led me to find this issue <https://github.com/thesofproject/avs-topology-xml/issues/16>.

And here another user solved his issue by configuring WirePlumber such that S16_LE is enforced for the DMIC.
So I checked how to do that and got some help by a friend. Here another unexpected hurdle to overcome was that Linux MINT is still using older WirePlumber 0.4.17 which we were not aware of. Of course we were first modifying `wireplumber.config` according to the new rules for WirePlumber 0.5.4 (older 0.4 versions are expecting Lua scripts in `~/.config/wireplumber/main.lua.d/`, but newer 0.5 versions expect SPA-JSON files in `~/.config/wireplumber/wireplumber.config.d/`).

Now - finally - that was missing last piece of the puzzle (and I say many thanks to the user filing the issue report). 

SUMMARY of cornerstones:
Needed was the use of
- New Intel AVS Driver (`snd_intel_dspcfg.dsp_driver=4`)
- Linux Kernel 6.8 (containing numerous fixes and subformat support)
- Alsa-Lib 1.2.11 (for subformat support)
- Alsa-Utils 1.2.11 (for subformat support)
- [Skylake firmware blobs](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/commit/?id=289e3a9777fd9aa9fb5ff721172bdb1f27e2d8c8) (included in Linux-Firmware 20230310)
- UCM tplg.bin files from <https://github.com/thesofproject/avs-topology-xml>
- Symlink for ALC255 topology (`hda-10ec0255-tplg.bin -> hda-generic-tplg.bin`)
- UCM config files from <https://github.com/alsa-project/alsa-ucm-conf/commit/45667605869bfc1d7d1305f7b9edd1ac8c5a43ae> (included in Alsa-UCM-Conf 1.2.13)
- ADDING an `avs_dmic.conf` in `/usr/share/alsa/ucm2/conf.d/avs_dmic` (renamed copy of `Acer-Lars-1.0.conf`)
    Here I think this should be added at <https://github.com/alsa-project/alsa-ucm-conf>
    to ensure continuous general roll-out
- Using PipeWire and MODIFYING `wireplumber.config` to enforce use of S16_LE for the avs-dmic node (see <https://pipewire.pages.freedesktop.org/wireplumber/daemon/configuration/alsa.html>)


Now the DMIC works with Skylake-H as it should - and I am so superhappy :-) !

Thank you both for your support and advice!
A special thanks is for you, dear @crojewsk, being active in this issue for such a long time!

Kind regards
Yours Gnoenk

P.S.: Next task is to optimize the sound-volume level...
